elastic/beats
# add dot_expand proceesor

## What does this PR do?

Adds a `dot_expand` processor utilizing a new `Expand` method on `MapStr` - the reverse of `Flatten`.

Given an input log line event: `{"foo.bar": 1}`, an event with expanded fields will be published:
```
{
  "@timestamp": "2020-08-06T22:10:36.197Z",
  "@metadata": {
    "beat": "filebeat",
    "type": "_doc",
    "version": "8.0.0"
  },
  "ecs": {
    "version": "1.5.0"
  },
  "host": {
    "os": {
      "name": "Mac OS X",
      "kernel": "19.4.0",
      "build": "19E287",
      "platform": "darwin",
      "version": "10.15.4",
      "family": "darwin"
    },
    ...
  },
  "foo": {
    "bar": 1
  }
}
```

## Why is it important?

See #17021

## Checklist


## Author's Checklist

- [ ] Performance of `MapStr.Expand`
- [ ] Overlapping/duplicated key expectations
- [ ] Additional edge cases

## How to test this PR locally

```
./filebeat -e -c /dev/null --strict.perms=false -E output.console.enabled=true -E filebeat.inputs='[{type: stdin, enabled:true, json.keys_under_root:true}]' -E processors='[{dot_expand:{}}]' -d '*'
```
paste some json in and see it expanded in the output document.

## Related issues

Closes #17021
